packages/lisk-p2p/src/peer_selector.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { NotEnoughPeersError } from './errors';
+import { Peer } from './peer';
+
+// For future extension of return type
+export interface PeerReturnType {
+	readonly peers: ReadonlyArray<Peer>;
+}
+export interface PeerOptions {
+	readonly [key: string]: string | number;
+}
+/* tslint:disable: readonly-keyword*/
+interface Histogram {
+	[key: number]: number;
+}
+interface HistogramValues {
+	height: number;
+	histogram: Histogram;
+	max: number;
+}
+/* tslint:enable: readonly-keyword */
+export const selectPeers = (
+	peers: ReadonlyArray<Peer>,
+	selectionParams: PeerOptions = { lastBlockHeight: 0 },
+	numOfPeers?: number,
+): PeerReturnType => {
+	const filteredPeers = peers.filter(
+		// Remove unreachable peers or heights below last block height
+		(peer: Peer) => peer.height >= selectionParams.lastBlockHeight,
+	);
+
+	if (filteredPeers.length === 0) {
+		return { peers: [] };
+	}
+
+	// Order peers by descending height
+	const sortedPeers = filteredPeers.sort((a, b) => b.height - a.height);
+
+	const aggregation = 2;
+
+	const calculatedHistogramValues = sortedPeers.reduce(
```

Comment:
I know that this is partly because of the rules about avoiding `for` loops but I find this hard to read.
Is it possible to simplify this a bit? I'll approve anyway if it's too tricky.

Reply:
@shuse2 What do you think about it? should I change to a `forLoop` but that will also make me to use `let` keyword

Reply:
@jondubois discussed with Shu and we removed some unnecessary initialization, and reduce will look a bit simpler now